src/main/java/net/mcreator/ui/workspace/resources/AbstractResourcePanel.java

```diff
@@ -0,0 +1,136 @@
+ *
+ * @param <T>
+ */
+public abstract class AbstractResourcePanel<T> extends JPanel implements IReloadableFilterable {
+
+	protected final WorkspacePanel workspacePanel;
+
+	private final TransparentToolBar bar = new TransparentToolBar();
+
+	protected final ResourceFilterModel<T> filterModel;
+	protected JSelectableList<T> elementList;
+
+	public AbstractResourcePanel(WorkspacePanel workspacePanel, ResourceFilterModel<T> filterModel,
+			ListCellRenderer<T> render) {
+		this(workspacePanel, filterModel, render, JList.VERTICAL);
+	}
+
+	public AbstractResourcePanel(WorkspacePanel workspacePanel, ResourceFilterModel<T> filterModel,
+			ListCellRenderer<T> render, int layoutOrientation) {
+		super(new BorderLayout());
+		setOpaque(false);
+
+		this.filterModel = filterModel;
+		this.workspacePanel = workspacePanel;
+
+		elementList = new JSelectableList<>(filterModel);
+		elementList.setOpaque(false);
+		elementList.setCellRenderer(render);
+		elementList.setSelectionMode(ListSelectionModel.MULTIPLE_INTERVAL_SELECTION);
+		elementList.setLayoutOrientation(layoutOrientation);
+		if (layoutOrientation == JList.HORIZONTAL_WRAP)
+			elementList.setVisibleRowCount(-1);
+
+		JScrollPane sp = new JScrollPane(elementList);
+		sp.setOpaque(false);
+		sp.setHorizontalScrollBarPolicy(ScrollPaneConstants.HORIZONTAL_SCROLLBAR_NEVER);
+		sp.getViewport().setOpaque(false);
+		sp.getVerticalScrollBar().setUnitIncrement(11);
+		sp.getVerticalScrollBar().setUI(new SlickDarkScrollBarUI((Color) UIManager.get("MCreatorLAF.DARK_ACCENT"),
+				(Color) UIManager.get("MCreatorLAF.LIGHT_ACCENT"), sp.getVerticalScrollBar()));
+		sp.getVerticalScrollBar().setPreferredSize(new Dimension(8, 0));
+
+		add("Center", sp);
+
+		elementList.addKeyListener(new KeyAdapter() {
+			@Override public void keyReleased(KeyEvent e) {
+				if (e.getKeyCode() == KeyEvent.VK_DELETE) {
+					deleteCurrentlySelected();
+				}
+			}
+		});
+
+		bar.setBorder(BorderFactory.createEmptyBorder(3, 5, 3, 0));
+
+		add("North", bar);
+	}
+
+	abstract void deleteCurrentlySelected();
+
+	@Override public void refilterElements() {
+		filterModel.refilter();
+	}
+
+	protected void addToolBarButton(String translationKey, ImageIcon icon, ActionListener actionListener) {
+		bar.add(createToolBarButton(translationKey, icon, actionListener));
+	}
+
+	protected void addToolBarButton(String translationKey, ImageIcon icon, MouseListener mouseListener) {
+		bar.add(createToolBarButton(translationKey, icon, mouseListener));
+	}
+
+	public static JButton createToolBarButton(String translationKey, ImageIcon icon, ActionListener actionListener) {
+		return createToolBarButton(translationKey, icon, actionListener, null);
+	}
+
+	public static JButton createToolBarButton(String translationKey, ImageIcon icon, MouseListener mouseListener) {
+		return createToolBarButton(translationKey, icon, null, mouseListener);
+	}
+
+	public static JButton createToolBarButton(String translationKey, ImageIcon icon,

```

Comment:
Any reason why these are public unlike the other two methods?

Reply:
![image](https://github.com/MCreator/MCreator/assets/16374228/5a223a0d-3f60-4304-ae24-b92fc0c71488)

Could be used in eg. variables panel, localization, and similar

Reply:
Hmm, you plan to settle them in another (more general-purpose) class later?

Reply:
Possibly, could be part of that other PR. Maybe in WorkspacePanel